File: snapshot/2009/src/rst/labs1/vb/Labs3.vb

' Full analytical model listed for *selected* structural elements
Public Class Lab3
    Implements IExternalCommand

    Public Function Execute(ByVal commandData As Autodesk.Revit.ExternalCommandData, ByRef message As String, ByVal elements As Autodesk.Revit.ElementSet) As Autodesk.Revit.IExternalCommand.Result Implements Autodesk.Revit.IExternalCommand.Execute
        Dim doc As Revit.Document = commandData.Application.ActiveDocument

        ' get cateories needed later in the loop
        Dim CatStruCols As Category = doc.Settings.Categories.Item(BuiltInCategory.OST_StructuralColumns)
        Dim CatStruFrmg As Category = doc.Settings.Categories.Item(BuiltInCategory.OST_StructuralFraming)
        Dim CatStruFndt As Category = doc.Settings.Categories.Item(BuiltInCategory.OST_StructuralFoundation)

        Dim sMsg As String = Nothing

        Dim iter As ElementSetIterator = doc.Selection.Elements.ForwardIterator
        While iter.MoveNext
            Dim elem As Revit.Element = iter.Current

            ' Structural WALL
            If TypeOf elem Is Wall Then
                Dim w As Wall = elem
                Try
                    Dim anaWall As AnalyticalModelWall = w.AnalyticalModel
                    If Not anaWall Is Nothing Then
                        If anaWall.Curves.Size > 0 Then
                            sMsg = "Analytical Model for Wall " & w.Id.Value.ToString & vbCrLf
                            GetAnalyticalModelWall(anaWall, sMsg)
                            MsgBox(sMsg)
                        End If
                    End If
                Catch
                End Try

                ' Strucural FLOOR
            ElseIf TypeOf elem Is Floor Then
                Dim f As Floor = elem
                Try
                    Dim anaFloor As AnalyticalModelFloor = f.AnalyticalModel
                    If Not anaFloor Is Nothing Then
                        If anaFloor.Curves.Size > 0 Then
                            sMsg = "Analytical Model for Floor " & f.Id.Value.ToString & vbCrLf
                            GetAnalyticalModelFloor(anaFloor, sMsg)
                            MsgBox(sMsg)
                        End If
                    End If
                Catch
                End Try

                ' Strucural CONTINUOUS FOOTING
            ElseIf TypeOf elem Is ContFooting Then
                Dim cf As ContFooting = elem
                Try
                    Dim ana3D As AnalyticalModel3D = cf.AnalyticalModel
                    If Not ana3D Is Nothing Then
                        If ana3D.Curves.Size > 0 Then
                            sMsg = "Analytical Model for Continuous Footing " & cf.Id.Value.ToString & vbCrLf
                            GetAnalyticalModelContFooting(ana3D, sMsg)
                            MsgBox(sMsg)
                        End If
                    End If
                Catch
                End Try

                ' One of Strucural Standard Families
            ElseIf TypeOf elem Is FamilyInstance Then
                Try
                    Dim fi As FamilyInstance = elem
                    Select Case fi.Category.Name

                        Case CatStruCols.Name ' "Structural Columns" for EN locale
                            Try
                                Dim anaFrame As AnalyticalModelFrame = fi.AnalyticalModel
                                If Not anaFrame Is Nothing Then
                                    sMsg = "Analytical Model for Structural Column " & fi.Id.Value.ToString & vbCrLf
                                    ListCurve(anaFrame.Curve, sMsg)
                                    ListRigidLinks(anaFrame, sMsg)
                                    ListSupportInfo(anaFrame.SupportData, sMsg)
                                    MsgBox(sMsg)
                                End If
                            Catch
                            End Try

                        Case CatStruFrmg.Name ' "Structural Framing" for EN locale
                            Try
                                Dim anaFrame As AnalyticalModelFrame = fi.AnalyticalModel
                                If Not anaFrame Is Nothing Then
                                    sMsg = "Analytical Model for Structural Framing " & _
                                            fi.StructuralType.ToString & " " & fi.Id.Value.ToString & vbCrLf
                                    ListCurve(anaFrame.Curve, sMsg)
                                    ListRigidLinks(anaFrame, sMsg)
                                    ListSupportInfo(anaFrame.SupportData, sMsg)
                                    MsgBox(sMsg)
                                End If
                            Catch
                            End Try

                        Case CatStruFndt.Name  ' "Structural Foundations" for EN locale
                            Try
                                Dim anaLoc As AnalyticalModelLocation = fi.AnalyticalModel
                                If Not anaLoc Is Nothing Then
                                    Dim pt As XYZ = anaLoc.Point
                                    sMsg = "Analytical Model for Foundation " & fi.Id.Value.ToString & vbCrLf & _
                                                       "  LOCATION = " & pt.X & ", " & pt.Y & ", " & pt.Z & vbCrLf
                                    ListSupportInfo(anaLoc.SupportData, sMsg)
                                    MsgBox(sMsg)
                                End If
                            Catch
                            End Try

                    End Select
                Catch
                End Try
            End If

        End While

        Return IExternalCommand.Result.Succeeded
    End Function

    Public Sub ListCurve(ByRef crv As Curve, ByRef s As String)

        If TypeOf crv Is Geometry.Line Then      'LINE

            Dim line As Line = crv
            Dim ptS As XYZ = line.EndPoint(0)
            Dim ptE As XYZ = line.EndPoint(1)

            s += "  LINE:" & ptS.X & ", " & ptS.Y & ", " & ptS.Z & " ; " & _
                             ptE.X & ", " & ptE.Y & ", " & ptE.Z & vbCrLf

        ElseIf TypeOf crv Is Geometry.Arc Then      'ARC

            Dim arc As Arc = crv
            Dim ptS As XYZ = arc.EndPoint(0)
            Dim ptE As XYZ = arc.EndPoint(1)
            Dim r As Double = arc.Radius

            s += "  ARC:" & ptS.X & ", " & ptS.Y & ", " & ptS.Z & " ; " & _
                            ptE.X & ", " & ptE.Y & ", " & ptE.Z & " ; R=" & r & vbCrLf

        Else        ' GENERIC PARAMETRIC CURVE
            If crv.IsBound Then
                s += "  BOUND CURVE " & crv.GetType.Name & " - Tessellated result:" & vbCrLf
                Dim pts As XYZArray = crv.Tessellate
                Dim pt As XYZ
                For Each pt In pts
                    s += "    PT:" & pt.X & ", " & pt.Y & ", " & pt.Z & " ; " & vbCrLf
                Next
            Else
                s += "  UNBOUND CURVE ??? - shouldn't ever be in an Analytical Model!" & vbCrLf
            End If

        End If

    End Sub

    Public Sub GetAnalyticalModelWall(ByRef anaWall As AnalyticalModelWall, ByRef s As String)
        Dim crv As Curve
        For Each crv In anaWall.Curves
            ListCurve(crv, s)
        Next
        ListSupportInfo(anaWall.SupportData, s)
    End Sub

    Public Sub GetAnalyticalModelFloor(ByRef anaFloor As AnalyticalModelFloor, ByRef s As String)
        Dim crv As Curve
        For Each crv In anaFloor.Curves
            ListCurve(crv, s)
        Next
        ListSupportInfo(anaFloor.SupportData, s)
    End Sub

    Public Sub GetAnalyticalModelContFooting(ByRef ana3d As AnalyticalModel3D, ByRef s As String)
        Dim crv As Curve
        For Each crv In ana3d.Curves
            ListCurve(crv, s)
        Next
        ListSupportInfo(ana3d.SupportData, s)
    End Sub

    ' New for RB3
    Public Sub ListSupportInfo(ByRef supData As AnalyticalSupportData, ByRef s As String)

        'If supData include valid support data.
        If (supData Is Nothing) Then
            s += "There is no support Data with this Element " & vbCrLf
        Else
            ' Supported or not?
            If supData.Supported Then
                s += vbCrLf & "Supported = YES, by:" & vbCrLf
            Else
                s += vbCrLf & "Supported = NO" & vbCrLf
                Return
            End If

            ' List all supports
            For Each supInfo As AnalyticalSupportInfo In supData.InfoArray
                Dim supEl As Revit.Element = supInfo.Element
                s += "  " & supInfo.SupportType.ToString() & " from elem.id=" & supEl.Id.Value.ToString & " cat=" & supEl.Category.Name & vbCrLf
            Next
        End If
    End Sub

    ' New for RB3
    Public Sub ListRigidLinks(ByRef anaFrm As AnalyticalModelFrame, ByRef s As String)

        s += vbCrLf & "Rigid Link START = "
        Dim rigidLinkStart As Curve = anaFrm.RigidLink(0)
        If rigidLinkStart Is Nothing Then
            s += "None" & vbCrLf
        Else
            ListCurve(rigidLinkStart, s)
        End If

        s += "Rigid Link END   = "
        Dim rigidLinkEnd As Curve = anaFrm.RigidLink(1)
        If rigidLinkEnd Is Nothing Then
            s += "None" & vbCrLf
        Else
            ListCurve(rigidLinkEnd, s)
        End If

    End Sub

End Class

File: snapshot/2009/src/rst/labs3/cs/Labs3.cs
#region Header
// Revit API .NET Labs
//
// Copyright (C) 2007-2008 by Autodesk, Inc.
//
// Permission to use, copy, modify, and distribute this software
// for any purpose and without fee is hereby granted, provided
// that the above copyright notice appears in all copies and
// that both that copyright notice and the limited warranty and
// restricted rights notice below appear in all supporting
// documentation.
//
// AUTODESK PROVIDES THIS PROGRAM "AS IS" AND WITH ALL FAULTS.
// AUTODESK SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTY OF
// MERCHANTABILITY OR FITNESS FOR A PARTICULAR USE.  AUTODESK, INC.
// DOES NOT WARRANT THAT THE OPERATION OF THE PROGRAM WILL BE
// UNINTERRUPTED OR ERROR FREE.
//
// Use, duplication, or disclosure by the U.S. Government is subject to
// restrictions set forth in FAR 52.227-19 (Commercial Computer
// Software - Restricted Rights) and DFAR 252.227-7013(c)(1)(ii)
// (Rights in Technical Data and Computer Software), as applicable.
#endregion // Header

#region Namespaces
using System;
using System.Collections;
using Autodesk;
using Autodesk.Revit;
using Autodesk.Revit.Elements;
using Autodesk.Revit.Geometry;
using Autodesk.Revit.Parameters;
using Autodesk.Revit.Structural;
using Autodesk.Revit.Utility;
using CmdResult = Autodesk.Revit.IExternalCommand.Result;
#endregion // Namespaces

namespace RstLabs
{
  // Full analytical model listed for *selected* structural elements
  public class Lab3 : IExternalCommand
  {
    public CmdResult Execute(
      ExternalCommandData commandData,
      ref string message,
      ElementSet elements )
    {
      Document doc = commandData.Application.ActiveDocument;
      Category CatStruCols = doc.Settings.Categories.get_Item(BuiltInCategory.OST_StructuralColumns);
      Category CatStruFrmg = doc.Settings.Categories.get_Item(BuiltInCategory.OST_StructuralFraming);
      Category CatStruFndt = doc.Settings.Categories.get_Item(BuiltInCategory.OST_StructuralFoundation);
      ElementSetIterator iter = doc.Selection.Elements.ForwardIterator();
      string sMsg = null;

      while (iter.MoveNext())
      {
        Autodesk.Revit.Element elem = (Autodesk.Revit.Element)iter.Current;
        // Structural WALL
        if (elem is Wall)
        {
          Wall w = (Wall)elem;
          try
          {
            AnalyticalModelWall anaWall = (AnalyticalModelWall)w.AnalyticalModel;
            if ((anaWall != null) && (anaWall.Curves.Size > 0))
            {
              sMsg = "Analytical Model for Wall " + w.Id.Value.ToString() + "\r\n";
              this.GetAnalyticalModelWall(ref anaWall, ref sMsg);
              RacUtils.InfoMsg(sMsg);
            }
          }
          catch
          {

          }
        }
        // Strucural FLOOR
        else if (elem is Floor)
        {
          Floor f = (Floor)elem;
          try
          {
            AnalyticalModelFloor anaFloor = (AnalyticalModelFloor)f.AnalyticalModel;
            if ((anaFloor != null) && (anaFloor.Curves.Size > 0))
            {
              sMsg = "Analytical Model for Floor " + f.Id.Value.ToString() + "\r\n";
              this.GetAnalyticalModelFloor(ref anaFloor, ref sMsg);
              RacUtils.InfoMsg(sMsg);
            }
          }
          catch
          {

          }
        }
        // Strucural CONTINUOUS FOOTING
        else if (elem is ContFooting)
        {
          ContFooting cf = (ContFooting)elem;
          try
          {
            AnalyticalModel3D ana3D = (AnalyticalModel3D)cf.AnalyticalModel;
            if ((ana3D != null) && (ana3D.Curves.Size > 0))
            {
              sMsg = "Analytical Model for Continuous Footing " + cf.Id.Value.ToString() + "\r\n";
              this.GetAnalyticalModelContFooting(ref ana3D, ref sMsg);
              RacUtils.InfoMsg(sMsg);
            }
          }
          catch
          {

          }
        }
        else if (elem is FamilyInstance)
        {
          try
          {
            FamilyInstance fi = (FamilyInstance)elem;
            string strCatName = fi.Category.Name;

            if (strCatName == CatStruCols.Name) // "Structural Columns" for EN locale
            {
              try
              {
                AnalyticalModelFrame anaFrame = (AnalyticalModelFrame)fi.AnalyticalModel;
                if (anaFrame != null)
                {
                  sMsg = "Analytical Model for Structural Column " + fi.Id.Value.ToString() + "\r\n";
                  Curve cur = anaFrame.Curve;
                  this.ListCurve(ref cur, ref sMsg);
                  this.ListRigidLinks(ref anaFrame, ref sMsg);
                  AnalyticalSupportData supportData = anaFrame.SupportData;
                  this.ListSupportInfo(ref supportData, ref sMsg);
                  RacUtils.InfoMsg(sMsg);
                }
              }
              catch
              {

              }
            }
            if (strCatName == CatStruFrmg.Name) // "Structural Framing" for EN locale
            {
              try
              {
                AnalyticalModelFrame anaFrame = (AnalyticalModelFrame)fi.AnalyticalModel;
                if (anaFrame != null)
                {
                  sMsg = "Analytical Model for Structural Framing " + fi.StructuralType.ToString()
                    + " " + fi.Id.Value.ToString() + "\r\n";
                  Curve cur = anaFrame.Curve;
                  this.ListCurve(ref cur, ref sMsg);
                  this.ListRigidLinks(ref anaFrame, ref sMsg);
                  AnalyticalSupportData supportData = anaFrame.SupportData;
                  this.ListSupportInfo(ref supportData, ref sMsg);
                  RacUtils.InfoMsg(sMsg);
                }
              }
              catch
              {

              }
            }
            if (strCatName == CatStruFndt.Name)  // "Structural Foundations" for EN locale
            {
              try
              {
                AnalyticalModelLocation anaLoc = (AnalyticalModelLocation)fi.AnalyticalModel;
                if (anaLoc != null)
                {
                  XYZ pt = anaLoc.Point;
                  sMsg = "Analytical Model for Foundation " + fi.Id.Value.ToString()
                    + "\r\n  LOCATION = " + pt.X.ToString() + ", " + pt.Y.ToString() + ", " + pt.Z.ToString()
                    + "\r\n";
                  AnalyticalSupportData supportData = anaLoc.SupportData;

                  this.ListSupportInfo(ref supportData, ref sMsg);
                  RacUtils.InfoMsg(sMsg);
                }
              }
              catch
              {

              }
            }
          }
          catch
          {

          }
        }
      }
      return CmdResult.Succeeded;
    }

    public void ListCurve(ref Curve crv, ref string s)
    {
      if ((crv is Autodesk.Revit.Geometry.Line))
      {
        // LINE
        Line line = crv as Autodesk.Revit.Geometry.Line;
        XYZ ptS = line.get_EndPoint(0);
        XYZ ptE = line.get_EndPoint(1);
        s = s + "  LINE:"
              + ptS.X + ", " + ptS.Y + ", " + ptS.Z + " ; "
              + ptE.X + ", " + ptE.Y + ", " + ptE.Z + "\r\n";
      }
      else if ((crv is Arc))
      {
        // ARC
        Arc arc = crv as Arc;
        XYZ ptS = arc.get_EndPoint(0);
        XYZ ptE = arc.get_EndPoint(1);
        double r = arc.Radius;
        s = s + "  ARC:"
              + ptS.X + ", " + ptS.Y + ", " + ptS.Z + " ; "
              + ptE.X + ", " + ptE.Y + ", " + ptE.Z
              + " ; R=" + r
              + "\r\n";
      }
      else
      {
        // GENERIC PARAMETRIC CURVE
        if (crv.IsBound)
        {
          s = s + "  BOUND CURVE " + crv.GetType().Name + " - Tessellated result:" + "\r\n";
          XYZArray pts = crv.Tessellate();

          foreach (XYZ pt in pts)
          {
            s = s + "  PT:"
                  + pt.X + ", " + pt.Y + ", "  + pt.Z + " ; "
                 + "\r\n";
          }
        }
        else
        {
          s = s + "  UNBOUND CURVE ??? - shouldn\'t ever be in an Analytical Model!" + "\r\n";
        }
      }
    }


    public void GetAnalyticalModelWall(ref AnalyticalModelWall anaWall, ref string s)
    {
      Curve crvTemp = null;
      foreach (Curve crv in anaWall.Curves)
      {
        crvTemp = crv;
        ListCurve(ref crvTemp, ref s);
      }
      AnalyticalSupportData supportData = anaWall.SupportData;
      ListSupportInfo(ref supportData, ref s);
    }

    public void GetAnalyticalModelFloor(ref AnalyticalModelFloor anaFloor, ref string s)
    {
      Curve crvTemp = null;
      foreach (Curve crv in anaFloor.Curves)
      {
        crvTemp = crv;
        ListCurve(ref crvTemp, ref s);
      }
      AnalyticalSupportData supportData = anaFloor.SupportData;
      ListSupportInfo(ref supportData, ref s);
    }

    public void GetAnalyticalModelContFooting(ref AnalyticalModel3D ana3d, ref string s)
    {
      Curve crvTemp = null;
      foreach (Curve crv in ana3d.Curves)
      {
        crvTemp = crv;
        ListCurve(ref crvTemp, ref s);
      }
      AnalyticalSupportData supportData = ana3d.SupportData;
      ListSupportInfo(ref supportData, ref s);
    }

    // New for RB3
    public void ListSupportInfo(ref AnalyticalSupportData supData, ref string s)
    {
      // If supData include valid support data.
      if (supData == null)
      {
        s = s + "There is no Support Data with this Element." + "\r\n";
      }
      else
      {
        // Supported or not?
        if (supData.Supported)
        {
          s = s + "\r\n" + "Supported = YES, by:" + "\r\n";
        }
        else
        {
          s = s + "\r\n" + "Supported = NO" + "\r\n";
          return;
        }
        // List all supports
        foreach (AnalyticalSupportInfo supInfo in supData.InfoArray)
        {
          Autodesk.Revit.Element supEl = supInfo.Element;
          s = s + "  " + supInfo.SupportType.ToString()
                 + " from elem.id=" + supEl.Id.Value.ToString()
                 + " cat=" + supEl.Category.Name
                 + "\r\n";
        }
      }
    }

    // New for RB3
    public void ListRigidLinks(ref AnalyticalModelFrame anaFrm, ref string s)
    {
      s = s + "\r\n" + "Rigid Link START = ";
      Curve rigidLinkStart = anaFrm.get_RigidLink(0);
      if (rigidLinkStart == null)
      {
        s = s + "None" + "\r\n";
      }
      else
      {
        ListCurve(ref rigidLinkStart, ref s);
      }
      s += "Rigid Link END   = ";
      Curve rigidLinkEnd = anaFrm.get_RigidLink(1);
      if ((rigidLinkEnd == null))
      {
        s = s + "None" + "\r\n";
      }
      else
      {
        ListCurve(ref rigidLinkEnd, ref s);
      }
    }
  }
}
